pkg/elemental/elemental.go

```diff
@@ -470,6 +471,109 @@ func DeployImage(c types.Config, img *types.Image) error {
 	return nil
 }
 
+// DeployRecoverySystem deploys kernel, initrd and image to the specified
+// paths.
+// This can be used for both ISO (all artifacts in same output dir) and raw
+// disks (kernel and initrd in ESP, rootfs squashfs image in recovery
+// partition.
+func DeployRecoverySystem(cfg types.Config, img *types.Image, bootDir string) error {
+	var err error
+	var cleaner func() error
+
+	cfg.Logger.Infof("Deploying recovery image: %s", img.File)
+	transientTree := strings.TrimSuffix(img.File, filepath.Ext(img.File)) + ".imgTree"
+	if img.Source.IsDir() {
+		transientTree = img.Source.Value()
+	} else if img.Source.IsFile() {
+		srcImg := &types.Image{
+			File:       img.Source.Value(),
+			MountPoint: transientTree,
+		}
+		err := MountFileSystemImage(cfg, srcImg)
+		if err != nil {
+			cfg.Logger.Errorf("failed mounting image tree: %v", err)
+			return err
+		}
+		cleaner = func() error {
+			err := UnmountFileSystemImage(cfg, srcImg)
+			if err != nil {
+				return err
+			}
+			return cfg.Fs.RemoveAll(transientTree)
+		}
+	} else if img.Source.IsImage() {
+		err = DumpSource(cfg, transientTree, img.Source)
+		if err != nil {
+			cfg.Logger.Errorf("failed dumping image tree: %v", err)
+			return err
+		}
+		cleaner = func() error { return cfg.Fs.RemoveAll(transientTree) }
+	}
+
+	kernel, initrd, err := utils.FindKernelInitrd(cfg.Fs, transientTree)
+	if err != nil {
+		cfg.Logger.Errorf("Could not find kernel and/or initrd: %s", err.Error())
+		return err
+	}
+
+	// Copy boot artifacts
+	for _, file := range []string{
+		kernel,
+		initrd,
+		filepath.Join(transientTree, cnst.GrubCfgPath, cnst.BootargsCfg),
+	} {
+		if exist, _ := utils.Exists(cfg.Fs, file); !exist {
+			cfg.Logger.Debugf("File '%s' not found, skipping", file)
+			continue
+		}
+
+		target := filepath.Join(bootDir, filepath.Base(file))
+		if exist, _ := utils.Exists(cfg.Fs, target); exist {
+			cfg.Logger.Debugf("Removing old file %s", target)
+			err = cfg.Fs.Remove(target)
```

Comment:
This is to make sure we deploy the newest files.

On error we could potentially leave the system unbootable because of a mix of kernel/initrd/grub-bootargs.

I think this is safe since we are updating from the active system, but we could consider deploying the artifacts into a temporary boot-dir and then moving it to `<recovery>/boot` once everything is deployed.

Reply:
Yes touching bootloader binaries is tricky. In fact this is also a valid concern we currently have for upgrades too when using the bootloader install flag. I'd be happy to brainstorm a little around this topic, if someone faces issues in that area, even it is unlikely, would be quite frustrating. I'd be in favor to apply some more elaborated logic, the question here is which would be the desired behavior.

Reply:
Agreed! I will create a ticket to stabilize the upgrade! It should probably include some integration tests as well.